### DISCUSSION TITLE
Improved (MediaWiki-like) definition list syntax

### DISCUSSION THREAD

**Discussion by bpj** [jgm/djot#193] [Ideas]
Below are some thought I have about definition lists. I have been a bit reluctant to come forward with them, but have decided to do so anyway in light of recent discussion in #35. It is also relevant to #192.

I think djot definition lists should take a hint from the MediaWiki definition list syntax, where both terms and definitions have an explicit marker:

``` mediawiki
; Ut
: Velit qui omnis delectus distinctio aliquid et dolor autem.
; Iusto
: Quae magnam quo laborum incidunt.
; Quod
: Mollitia nostrum magnam incidunt nesciunt at.
```

Transferred to djot, where multi-line definitions and nested lists should be possible this becomes something like

``` djot
; title

: Definition list syntax

; author 

: - ; name

    : Iusto Veritatis

    ; email

    : aut@example.org
    
    : nihil@similique.example.com

  - ; name

    : Optio Blanditiis

    ; email

    : voluptatibus@example.com

  - ; name

    : Vel Dolore

    ; email

    : ratione@example.net
    
; date:

: ; 2022-12-18

  : First version.
  
  ; 2023-01-15
  
  : Added some wild ideas.
```

I think this is preferable because

- it becomes immediately clear which is a term and which is a definition because both have a specific marker,

- it should be easier to parse, I think.

- not least it allows for more than one definition per term, which is a possibility I would like to have.

- the term and definition(s) being at the same indentation level reduces indentitis in nested lists (although this is a minor concern!)

Of course djot might want to use different markers, or reverse them, to preserve some backwards similarity but I think the colon at the start of a definition and the semicolon at the start of the next term makes a nice pair.

## A possible compact definition list syntax?

I’ll venture out on a limb here; what follows is nowise an essential part of this proposal, just a possibile extension with (I think) still simple rules. Note that I still have my doubts about compactness as a goal in itself!

If there is a rule that terms may not contain any unescaped colons (or: no unescaped colons with whitespace on both sides!) a “tight” one-line syntax becomes possible e.g. in metadata (#35):

``` djot
:::meta
; title : Perferendis adipisci aut
; author : est@example.com
; date : 2023-01-15
:::
```

but it should of course be possible to mix this with the multiline/spaced syntax! The rule would be that a semicolon at the start of a line, with/without indentation, would mark the start of a term, and any following inlines not containing any unescaped colon (but presumably allowing `{:symb:}` and inline verbatim) would be the term. If the term is followed by `/[ \t]+\:[ \t]+/` on the same line inlines up to the next blank line are the (only) definition. If the term is followed by a blank line and the next line starts with an unescaped colon on the same indentation level everything, including indented blocks, up to a block starting at the same indentation level is a definition. It may be followed by more such definitions up to another term on the same indentation level, or a block not starting with a colon/semicolon on the same indentation level, which will be the first block after the definition list:

``` djot
The block before a definition list.

; term : one-liner definition.
; A term with\: `:stuff`!  {:grin:}
: this is also a one-liner def; actually
; This is a term to{:exclamation:}

: This is a block definition.

: This is the next block def under the same term.

: ; Definition list

  : Nested inside the third definition of the outer list.
  
    The first definition in the nested definition list has multiple blocks.
    
  ; The second item of the nested definition list : It is a oneliner.
  
; Another term in the outer definition list

: - Its definition.
  - It contains a bullet list.
  
The block after the definition list.
```

**Comment by bpj** [comment]
BTW definition lists should probably be renamed e.g. "key-value list"  or "itemize list" since in my experience they are more often used for other things than strictly definitions of terms/glossaries.

**Reply by dimztimz** [reply]
In HTML 5 it is called description list, it got renamed from the HTML 4 definition list.

**Comment by jgm** [comment]
These are some interesting ideas. I can see the use of having `;` to mark a term and `:` a definition. It would allow multiple definitions, and it would allow a more compact form (even if we didn't go for your one-line idea):
```
; term
: definition
```

But I'm curious what you think about reStructuredText's idea of having two distinct concepts: definition lists and field lists.  Note that field lists can contain block-level content via indentation.

In djot, a field list would, essentially, ust be a list where the markers are distinct field names.

```
:title: The title
:abstract: Put whatever you like here

  - even
  - a list
:editor:
  :given: S. J.
  :family: Jones
```

These wouldn't replace definition lists (because of the restriction on field names); it would be a distinct concept; but these seem better suited for metadata.

**Reply by bpj** [reply]
Do you mean that djot would use that exact syntax for field lists? I sense ambiguity with symb syntax!

**Comment by jgm** [comment]
If we used the same syntax we might need to rethink symb.  But I was really wondering more about the concept than the syntax.

**Reply by matklad** [reply]
What would be the semantics? Presumably definition list is the real definition list, and field list is for meta, which doesn’t have any *specific* display representation (it’s up to converter to interpret meta, default converter might just omit it from output, or display in an unspecified way)?

To me it does seem worthwhile to syntactically distinguish the two. While we can re-use definition lists, I think this might be confusing for users.

syntax-wise, both asciidoctor and restructured text use `:key: value` concrete syntax. They however have some pre-defined attributes interpreted by the tool itself. That bit I’d rather we avoid in djot.

**Comment by jgm** [comment]
reST uses field lists in two main contexts:

- after the title, where it's interpreted as metadata
- as a way to add attributes to directives, e.g. images (from memory):

```
.. image:: foo.jpg
  :title: Bar
```

We don't need this for attributes, so there's the question how a field list would render outside of a metadata context.  In rST a field list that isn't metadata or attributes renders as a table with class field-list.

One option would be to do what pandoc does and allow metadata fields to be specified anywhere in the document; in this case, they'd just accumulate.

**Comment by uvtc** [comment]
The `;` and `:` are extremely similar looking. I think it would decrease readability to have them together where the reader needs to tell them apart.

If it makes sense to use two markers for def list item and def list definition, the first thing that comes to my mind is:

~~~
apple:

 = Grows on trees. Pick them in the autumn.

orange:

 = Round. Comes in an orange color.

pineapple:

 = No pine involved here. Nor apple for that matter.
~~~

And here that is in a tighter format:

~~~
apple:
 = Grows on trees. Pick them in the autumn.
orange:
 = Round. Comes in an orange color.
pineapple:
 = No pine involved here. Nor apple for that matter.
~~~

where the spaces around the `=` are required for two reasons: (1) to disambiguate it from `{=highlight=}`, and (2) for readability.

**Comment by tbdalgaard** [comment]
I like this. It is easy to remember, read and write.



Sendt fra min Mac Mini via Apple Mail

**Comment by alxlg** [comment]
I had a similar requirement for Logseq that currently uses Markdown + a custom syntax for key-value pairs.

I recently suggested to it the following Markdown-ish syntax that could be adapted to Djot if you are interested:

https://discuss.logseq.com/t/an-idea-for-a-more-standard-markdown-property-syntax/20073

Example:

```
- Oscar Wilde
  * Description:: Oscar Wilde was an Irish poet and playwright. After writing in different forms throughout the 1880s, he became one of the most popular playwrights in London in the early 1890s.
  * Books::
    - The Picture of Dorian Gray
      * Year:: 1891
	- The Importance of Being Earnest
	  * Year:: 1895
- Lewis Carroll
  * Books:: [[Alice in Wonderland]]
```

Rendered:

- Oscar Wilde
  * Description:: Oscar Wilde was an Irish poet and playwright. After writing in different forms throughout the 1880s, he became one of the most popular playwrights in London in the early 1890s.
  * Books::
    - The Picture of Dorian Gray
      * Year:: 1891
	- The Importance of Being Earnest
	  * Year:: 1895
- Lewis Carroll
  * Books:: [[Alice in Wonderland]]

**Comment by paradox460** [comment]
AsciiDoc has a different take on Definition lists (which they call description lists)

https://docs.asciidoctor.org/asciidoc/latest/lists/description/

Basically, you write them like this

```adoc

Key:: Value
Key 2:: Value 2
Key with children::
  child::: This is a child
child::: This is also a child, note that there are 3 colons; indentation does not matter
```

Might not fit with the rest of Djot, given its a "repeating character" based delimiter, but it does provide a pretty convenient interface for writing lists.

Interestingly enough, Kramdown and PHPMarkdownExtra have an almost inverted approach to how Djot handles definition lists

```kramdown
Key
: value
Key 2
: value 2
```

**Reply by jgm** [reply]
Yes, that PHP Markdown Extra behavior is used in many markdown variants, including pandoc. A disadvantage is that there's no marker at the beginning of a block that lets the parser know we've got a definition list.

**Comment by itraveller1** [comment]
I fully agree with the original idea. I would just like to make a suggestion regarding the labeling.

~~~
@ term
: description one
: description two
~~~

The "@" symbol is quite expressive and relatively rare in regular text. Even its basic semantics "at" is very appropriate for our case - what the descriptions refer to.

The biggest plus is that the proposed structure is completely equivalent to a regular list. By the way, descriptions without a term are also very popular as explanations and generalizations.

**Reply by jgm** [reply]
> By the way, descriptions without a term are also very popular as explanations and generalizations.

Examples?

**Reply by itraveller1** [reply]
Sorry, I meant similar cases in extended Markdown, where a construction like `: ...` at the beginning of a line is sometimes used for indentation. For example, when a description is needed not for a single word or phrase, but for a whole block or section. The phrase "very popular" is clearly unnecessary here.

I know that in HTML it is incorrect to use a list of descriptions without a term, but HTML, unlike Markdown, has enough other means of style design. In regular text, indentation is used not only to separate paragraphs - that's why users resort to this kind of "violation".

A block of quotes is also a bad replacement, since they have their own semantics, completely different from descriptions. Perhaps such construction should have its own interpretation?

**Reply by jgm** [reply]
I don't think semantic constructions like definite lists should be abused just to get an indented paragraph.  If you want an indented paragraph, and you don't want it to be a block quote, then maybe attach an attribute to it, or put it in a special div with an attribute, and style that using CSS or a filter.
